subsys/usb/class/hid/core.c

```diff
@@ -149,11 +149,30 @@ static void hid_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
 	hid_device.ops->int_in_ready();
 }
 
+#ifdef CONFIG_ENABLE_HID_INT_OUT_EP
+static void hid_int_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+{
+	if (hid_device.ops->int_out_read  &&
+		(ep_status == USB_DC_EP_DATA_OUT)) {
+
+		u8_t data[CONFIG_HID_INTERRUPT_EP_MPS];
+		u32_t len = 0;
+
+		usb_read(ep, data, sizeof(data), &len);
```

Comment:
I have tried that implementation. I was not able receive the correct data sent from host. so I had put read in callback and added application call back.

Reply:
What do you mean by "I was not able receive the correct data"? The data was corrupted? the wrong length?

I have just tried locally and it works perfectly. Here is the patch I used on top of your patchset:

```
--- a/include/usb/class/usb_hid.h
+++ b/include/usb/class/usb_hid.h
@@ -54,9 +54,6 @@ struct usb_hid_descriptor {
 typedef int (*hid_cb_t)(struct usb_setup_packet *setup, s32_t *len,
 			u8_t **data);
 typedef void (*hid_int_ready_callback)(void);
-#ifdef CONFIG_ENABLE_HID_INT_OUT_EP
-typedef void (*hid_out_ready_callback)(void *data, u32_t len);
-#endif
 
 struct hid_ops {
 	hid_cb_t get_report;
@@ -73,7 +70,7 @@ struct hid_ops {
 	 */
 	hid_int_ready_callback int_in_ready;
 #ifdef CONFIG_ENABLE_HID_INT_OUT_EP
-	hid_out_ready_callback int_out_read;
+	hid_int_ready_callback int_out_ready;
 #endif
 };
 
--- a/subsys/usb/class/hid/core.c
+++ b/subsys/usb/class/hid/core.c
@@ -152,15 +152,11 @@ static void hid_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
 #ifdef CONFIG_ENABLE_HID_INT_OUT_EP
 static void hid_int_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
 {
-	if (hid_device.ops->int_out_read  &&
-		(ep_status == USB_DC_EP_DATA_OUT)) {
-
-		u8_t data[CONFIG_HID_INTERRUPT_EP_MPS];
-		u32_t len = 0;
-
-		usb_read(ep, data, sizeof(data), &len);
-		hid_device.ops->int_out_read(data, len);
+	if (ep_status != USB_DC_EP_DATA_OUT ||
+	    hid_device.ops->int_out_ready == NULL) {
+		return;
 	}
+	hid_device.ops->int_out_ready();
 }
#endif
```


Then, here is the way I read the data in the application:

```
void hid_out()
{
        u8_t data[64];
        u32_t len = 0;

        usb_read(CONFIG_HID_INT_OUT_EP_ADDR, data, sizeof(data), &len);

        /* use the data */
}
```

Reply:
data was corrupted. I have also tested. My test target used usb_dw driver.

Reply:
For me I tested it on the usb_stm32 driver. That said I don't fully get it, doing the read from the callback is just doing it a few instructions after and it's still done in the workqueue thread. It's only done one function call after, could it be that your workqueue stack is too small?

Reply:
One of the difference of doing a callback function is that the call to usb_read uses the `CONFIG_HID_INT_OUT_EP_ADDR` endpoint instead of the `ep` endpoint passed to hid_int_out. Could it happen that they don't match?

Also you might want to use a different IN and OUT endpoint addresses to check that the DW driver doesn't mix data from both endpoints.